File: src/CSharp/238.ProductOfArrayExceptSelf.cs

using System;
using System.Collections.Generic;

namespace PeefyLeetCode.ProductOfArrayExceptSelf
{

public class Solution {
    public int[] ProductExceptSelf(int[] nums) {
        int[] res = new int[nums.Length];
        int p = 1, q = 1;
        for (int i = 0; i < nums.Length; i++) {
            res[i] = p;
            p *= nums[i];
        }
        for (int i = nums.Length - 1; i > 0 ; i--) {
            q *= nums[i];
            res[i - 1] *= q;
        }
        return res;
    }
}
}

File: src/Python/238.ProductOfArrayExceptSelf.py

class Solution:
    def productExceptSelf(self, nums):
        n = len(nums)
        left = 1
        right = 1
        res = [1] * n
        for i in range(n):
            res[i] *= left
            left *= nums[i]
            res[n - 1 - i] *= right
            right *= nums[n - 1 - i]
        return res


if __name__ == "__main__":
    solution = Solution()
    print(solution.productExceptSelf([1, 2, 3, 4]))
    
